### ISSUE TITLE
Binary TextLayer With Font Icons?

### ISSUE THREAD

**Issue by jfuehner** [visgl/deck.gl#4434]
Would there be an example of using the deck 8.1 [TextLayer ](https://deck.gl/#/documentation/deckgl-api-reference/layers/text-layer) binary attributes with a font icon?

Previous to the available binary implementation for the TextLayer I was able to pass in `characterSet`. Doing the same with the binary attributes the font icon's are not rendering as expected. I assume it has to do with passing the code points in for the `getText` attribute.

As always a nudge in the correct direction would be greatly appreciated.

**Comment by Pessimistress** [comment]
> Doing the same with the binary attributes the font icon's are not rendering as expected

Please include a screenshot/code sample when reporting issues.

**Comment by jfuehner** [comment]
@Pessimistress the docs do not speak to passing in the `characterSet` differently for binary. Does the characterSet need to be in the same `charCode`/code point format?

**Comment by Pessimistress** [comment]
You should still pass `characterSet` when using binary.

**Comment by jfuehner** [comment]
@Pessimistress I created this [CodePen](https://codepen.io/jfuehner/pen/LYVobRL) which shows the issue I'm having passing in the binary representation of the unicode string in conjunction with the `characterSet` for icon [F0643](https://cdn.materialdesignicons.com/5.0.45/).

Deck displays a warning message indicating the characters are missing. I received this warning previously when working with the non-binary attributes but found that I wasn't escaping the `\u` properly.

With the binary attributes I have tried escaping with single and double-backslashes but still am not able to get deck to resolve the icon.

Thoughts?

**Comment by Pessimistress** [comment]
Change your code to

```js
const TEXT_DATA = [
  {
    text: '\u2190\u2191\u2192',
    position: [-122.4, 37.7],
    color: [255, 0, 0]
  }
];

const texts = new Uint16Array(...);
```

And

```js
characterSet: ['\u2190','\u2191','\u2192'],
```

- An easy way to test your character escape is to simply `console.log` your text string.
- Check the char code range by calling `str.charCodeAt(i)` - `Uint8Array` handles values up to 255. `Uint16Array` handles values up to 65535. If you are using arbitrary unicode characters you need `Uint32Array`.

**Comment by jfuehner** [comment]
Thanks! I updated the [CodePen](https://codepen.io/jfuehner/pen/LYVobRL) and do see that the `TextLayer` is working with the built-in unicode characters. Why do you suppose the unicode values from the custom [font icon set](https://cdn.materialdesignicons.com/5.0.45) are not working as expected?

**Comment by Pessimistress** [comment]
The current implementation does not handle unicode that is outside of the UTF-16 range. I'll put up a patch.

**Comment by Pessimistress** [comment]
@jfuehner Note that when the above PR is merged, you still need to make the following changes to your code:

- `'\u{F0643}'` instead of `'\uF0643'` (read about the syntax [here](https://flaviocopes.com/javascript-unicode/))
- Use [codePointAt](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/codePointAt) instead of `charCodeAt`
- Use `Uint32Array` to hold the character codes as explained in my previous comment

**Comment by jfuehner** [comment]
@Pessimistress, I made the change you committed [d29ba8d](https://github.com/uber/deck.gl/commit/d29ba8d58b52dbb59ef5b1fa669e1d83710feafb) manually to my local version of deck `text-layer/utils`.  

I then modified my code to use `codePointAt` as you suggested...

```
          const texts = new Uint32Array(TEXT_DATA.map(d => {
              return Array.from(d.text).map(_ => _.codePointAt(0));
            }).flat());

```
```
                new TextLayer({
                  characterSet: ['\u{F0643}'],
                  fontFamily: 'Material Design Icons',
                  data: {
                    length: TEXT_DATA.length,
                    startIndices: startIndices,
                    attributes: {
                      getText: {value: texts},
                      getPosition: {value: positions, size: 2},
                      getColor: {value: colors, size: 3}
                    }
                  }
                })
```

However, this still resulted in deck reporting the missing character warning...

![image](https://user-images.githubusercontent.com/1843941/79283041-aa875480-7e84-11ea-87d1-4c168dcf492d.png)

**Comment by Pessimistress** [comment]
Fix published in 8.1.3.

**Comment by enodr** [comment]
Hi,

I would be interested in a working example for the above issue. I am trying to use also a Font Icon (Material Design Icons) but cannot solve the characterSet issue. The abode codepen example is not working even with the suggested changes.
I almost always get the "deck: Missing character" error.
